From e3ca887f7d8a95fcf0dfaa03291a7f9740b1f572 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:12:02 -0400
Subject: [powerpc] powernv: Missed IOMMU table type

Message-id: <1410545655-205645-494-git-send-email-dzickus@redhat.com>
Patchwork-id: 94389
O-Subject: [RHEL7 PATCH 493/626] powerpc/powernv: Missed IOMMU table type
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 65fd766b990f79c9b541050a3f868ad058ad0d69
Author: Gavin Shan <gwshan@linux.vnet.ibm.com>
Date:   Thu Apr 24 18:00:28 2014 +1000

    powerpc/powernv: Missed IOMMU table type

    In function pnv_pci_ioda2_setup_dma_pe(), the IOMMU table type is
    set to (TCE_PCI_SWINV_CREATE | TCE_PCI_SWINV_FREE) unconditionally.
    It was just set to TCE_PCI by pnv_pci_setup_iommu_table(). So the
    primary IOMMU table type (TCE_PCI) is lost. The patch fixes it.

    Also, pnv_pci_setup_iommu_table() already set "tbl->it_busno" to
    zero and we needn't do it again. The patch removes the redundant
    assignment.

    The patch also fixes similar issues in pnv_pci_ioda_setup_dma_pe().

    Signed-off-by: Gavin Shan <gwshan@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/platforms/powernv/pci-ioda.c b/arch/powerpc/platforms/powernv/pci-ioda.c
index 9f801c0..30888a4 100644
--- a/arch/powerpc/platforms/powernv/pci-ioda.c
+++ b/arch/powerpc/platforms/powernv/pci-ioda.c
@@ -643,10 +643,10 @@ static void pnv_pci_ioda_setup_dma_pe(struct pnv_phb *phb,
    * errors, and on the first pass the data will be a relative
    * bus number, print that out instead.
    */
-  tbl->it_busno = 0;
   tbl->it_index = (unsigned long)ioremap(be64_to_cpup(swinvp), 8);
-  tbl->it_type = TCE_PCI_SWINV_CREATE | TCE_PCI_SWINV_FREE |
-          TCE_PCI_SWINV_PAIR;
+  tbl->it_type |= (TCE_PCI_SWINV_CREATE |
+     TCE_PCI_SWINV_FREE   |
+     TCE_PCI_SWINV_PAIR);
  }
  iommu_init_table(tbl, phb->hose->node);
  iommu_register_group(tbl, pci_domain_nr(pe->pbus), pe->pe_number);
@@ -771,9 +771,8 @@ static void pnv_pci_ioda2_setup_dma_pe(struct pnv_phb *phb,
    * errors, and on the first pass the data will be a relative
    * bus number, print that out instead.
    */
-  tbl->it_busno = 0;
   tbl->it_index = (unsigned long)ioremap(be64_to_cpup(swinvp), 8);
-  tbl->it_type = TCE_PCI_SWINV_CREATE | TCE_PCI_SWINV_FREE;
+  tbl->it_type |= (TCE_PCI_SWINV_CREATE | TCE_PCI_SWINV_FREE);
  }
  iommu_init_table(tbl, phb->hose->node);
 
-- 
1.7.1